```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      eval = FALSE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# General Troubleshooting {#general-troubleshooting}

<style>
body {
text-align: justify}
</style>

<span style="color:red;">Important Note!!</span> This handbook was written exclusively for people working at the RIVM (especially people at IDS). Some of the parameters or the way to run the pipeline would need to be changed for external people. Refer to the README section on the GitHub page of the corresponding pipeline. The pipelines that are enlisted as available through the internal GitLab from the RIVM are not accessible for external people. This might change in the future but for now that is the case. Equally, although the team would love to help external people to use our pipelines, we cannot give the personalized and fast-responding help that we give to people at the RIVM. If you do need help please write an Issue in the corresponding GitHub repository and we will try to help as soon as possible. 

## Setting up conda and avoiding/solving problems when installing/activating a pipeline environment {#set-conda}

If you are not used to `conda` environments yet, you may not have your `conda` software completely set-up. This can cause, among other things, that you get a message like this when you are trying to follow the instructions of one of our pipelines:

```
CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>
```

What you need to configure `conda` and fix this issue is:  

1. Go to the scratch_dir partition and make a folder with your RIVM user name. Replace <username> by your actual username (the RIVM one).

```
cd /mnt/scratch_dir
mkdir -p <username>
```

2. Then you can go inside the folder you just created and make extra folders where your `conda` environments will be stored:

```
cd /mnt/scratch_dir/<username>/
mkdir -p conda
mkdir -p conda/envs
mkdir -p conda/pkgs
```

3. The next step is a bit more difficult. First you move back to your home directory, then you "initiate" `conda` and finally you give it the paths where you want it to make your environments (the folders you just made in scratchdir):

```
cd /home/<username>/
conda init bash
touch /home/<username>/.condarc
```

To add the paths you can do it from the command line using the `vim .condarc` editor if you are familiar with how `vim` works. If you don't know how to do that, you can simply go to your home directory by going to the menu "Places" > "Home" on the top left of your Linux desktop. You can see the some files/folders that you have on your desktop but the .condarc file might be hidden. To show it you can go to the menu (three parallel lines on the top right of the 'Places' window) and make sure to have marked the 'Show Hidden Files' option. You should now be able to see the .condarc file listed. Please open it by double-clicking the file name. Finally, add or replace the following text with the correct paths (replace <username> with your RIVM user name, meaning the folders you just made in scratch_dir):

```
pkgs_dirs:
  - /mnt/scratch_dir/<username>/conda/pkgs
envs_dirs:
  - /mnt/scratch_dir/<username>/conda/envs
channel_priority: disabled
```

Now you are ready to make your conda environments. If you already tried making some environments before following these configuration steps, you may get an error. Specifically, if you try to use `conda activate <env_name>` you may get this error message:

```
Could not find conda environment: <env_name>
You can list all discoverable environments with `conda info --envs`.
```

Then you can try to re-install the pipeline environment following the pipeline instructions. If the problem persists, you probably already made a `conda` environment but in the wrong place. In that case we need to find the wrong one and delete it. For that please type:

```
conda env list
```

A list of your current environments will be displayed and will look something like this:

```
# conda environments:
#
base                  *  /mnt/miniconda
                         /home/<username>/.conda/envs/<env_pipeline>
mamba                    /home/<username>/.conda/envs/mamba
```

DO NOT DO ANYTHING WITH THE `base` environment. However, you do need to delete the other environments that were made in the wrong folder (in this case the `/home/<username>/.conda` folder instead of the `/mnt/scratch_dir/<username>/conda` folder). You can do it by typing:

```
rm -rf /home/<username>/.conda
```
**Note: Replace `/home/<username>/.conda` by the conda folder where the environments were meade. MAKE SURE TO ONLY DELETE THE `.conda` FOLDER. YOU SHOULD NEVER DELETE YOUR WHOLE HOME DIRECTORY.** Please ask a bioinformatician for help if you are unsure how to do this step.

This step might take a while. After this step is done, you can try to install the pipeline again (making the new environments)according to the pipeline instructions. If everything went fine, the `conda activate <pipeline_env>` command should work fine. If you wonder how do you know whether this command worked or not, you can just look at your command line.  

Before activating an environment it should look like this:

```
(base) [hernanda@rivm-biohn-l05p rondzendingen]$
```
* Look at the word `(base)` at the beginning of the line.  

After running the `conda activate <pipeline_env>` command it should look like this:

```
(<pipeline_env>) [hernanda@rivm-biohn-l05p rondzendingen]$ 
```
* The name of the environment you just activated should now appear at the beginning of the line `(<pipeline_env>)`.  

Also, when you do `conda env list` you should see the correct paths for your environments:

```
# conda environments:
#
base                  *  /mnt/miniconda
<env_pipeline>           /mnt/scratch_dir/<username>/.conda/envs/<env_pipeline>
mamba                    /mnt/scratch_dir/<username>/.conda/envs/mamba
```

## Failure when installing master environment  

Sometimes the installation does not work because of the settings of Linux. The installation always takes some minutes in which a message saying `Solving environment...` might appear on the screen. However, if this step takes too long (for instance 30 min or more) please abort the step by pressing `Ctrl + C` and/or `Ctrl + Z` (sometimes you need to press repeateadly). In such cases, please change the necessary settings with this command:  

```{bash}
conda config --set channel_priority false
```

Then try to run the pipeline and install the master environment again. 

## Failure to make a sample sheet or to find input directory   

There are few reasons why this could happen:

- Your path (all the folders that lead to your input directory) contains spaces or unrecognized characters.  

- You mis-spelled something in your path or gave the wrong path. Especially while working in Linux, it is common to forget a slash ("/") at the beginning of a full path. For instance: `/mnt/scratch_dir/<my_folder>/<my_data>/`.  

- Your files might have the incorrect format and/or extension. Normally for all the files, standard names are accepted (for instance, .fastq.gz or .fastq for raw reads and .fasta for fasta files). Please make sure that your files have the correct and "classical" extensions.   

- Your input directory does nor contain the expected files, but they are instead in sub-folders. Most pipelines expect all input files to be together in one folder instead of scattered in sub-folders. The only exception is when you use output from a previous pipeline (for instance, Juno-assembly) into another pipeline. Please collect all your samples together in one folder.  

## Other problems or failing rules  

The Juno pipelines are still in development which means that sometimes things fail. **Before contacting me** for help, please try these steps:  

1. Re-run the pipeline again and see if it goes further. If it does, please keep re-running the pipeline until your analysis is finished or it just doesn't go further. Even if you are able to finish your analysis, just send me an email afterwards (see step 3) so I can check what happened.    
2. Download the pipeline again and start from the beginning of this handbook. Sometimes the issue has been resolved in newer versions of the pipeline.
3. Collect your logging files and contact me. Please inform me about bugs/errors via [e-mail](mailto:alejandra.hernandez.segura@rivm.nl) **sending also your `log` files and the path where I can find your input directory and the pipeline**. No screenshots are necessary. Note that if you do not send this information, I will not be able to help you and your and my work will be delayed.  
