<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Juno-assembly pipeline | Bioinformatics Pipelines for Bacteriology (IDS, RIVM)</title>
  <meta name="description" content="3 Juno-assembly pipeline | Bioinformatics Pipelines for Bacteriology (IDS, RIVM)" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Juno-assembly pipeline | Bioinformatics Pipelines for Bacteriology (IDS, RIVM)" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Juno-assembly pipeline | Bioinformatics Pipelines for Bacteriology (IDS, RIVM)" />
  
  
  

<meta name="author" content="Alejandra Hernandez-Segura" />


<meta name="date" content="2021-10-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="general-troubleshooting.html"/>
<link rel="next" href="juno-typing.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="custom_style.css" type="text/css" />
<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics Pipelines for Bacteriology (IDS, RIVM)</a></li>  |
<li class="toc-logo"><a href="./"><img src="figures/juno_logo/juno_light/juno_bgp_lightbg.png" width="100%"></a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction and General Instructions</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#general-instructions"><i class="fa fa-check"></i><b>1.1</b> General Instructions for all pipelines</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#set-conda"><i class="fa fa-check"></i><b>1.1.1</b> Setting up conda</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#requirements-and-preparation"><i class="fa fa-check"></i><b>1.1.2</b> Requirements and preparation</a></li>
<li class="chapter" data-level="1.1.3" data-path="index.html"><a href="index.html#downloading-instructions"><i class="fa fa-check"></i><b>1.1.3</b> Downloading pipelines</a></li>
<li class="chapter" data-level="1.1.4" data-path="index.html"><a href="index.html#what-to-expect"><i class="fa fa-check"></i><b>1.1.4</b> What to expect while running a Juno pipeline</a></li>
<li class="chapter" data-level="1.1.5" data-path="index.html"><a href="index.html#general-output"><i class="fa fa-check"></i><b>1.1.5</b> General output and log files for every pipeline</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>I General troubleshooting</b></span></li>
<li class="chapter" data-level="2" data-path="general-troubleshooting.html"><a href="general-troubleshooting.html"><i class="fa fa-check"></i><b>2</b> General Troubleshooting</a>
<ul>
<li class="chapter" data-level="2.1" data-path="general-troubleshooting.html"><a href="general-troubleshooting.html#problems-conda-activate"><i class="fa fa-check"></i><b>2.1</b> Solving problems when installing a pipeline or activating a pipeline environment</a></li>
<li class="chapter" data-level="2.2" data-path="general-troubleshooting.html"><a href="general-troubleshooting.html#failure-when-installing-master-environment"><i class="fa fa-check"></i><b>2.2</b> Failure when installing master environment</a></li>
<li class="chapter" data-level="2.3" data-path="general-troubleshooting.html"><a href="general-troubleshooting.html#failure-to-make-a-sample-sheet-or-to-find-input-directory"><i class="fa fa-check"></i><b>2.3</b> Failure to make a sample sheet or to find input directory</a></li>
<li class="chapter" data-level="2.4" data-path="general-troubleshooting.html"><a href="general-troubleshooting.html#other-problems-or-failing-rules"><i class="fa fa-check"></i><b>2.4</b> Other problems or failing rules</a></li>
</ul></li>
<li class="part"><span><b>II Juno pipelines</b></span></li>
<li class="chapter" data-level="3" data-path="juno-assembly.html"><a href="juno-assembly.html"><i class="fa fa-check"></i><b>3</b> Juno-assembly pipeline</a>
<ul>
<li class="chapter" data-level="3.1" data-path="juno-assembly.html"><a href="juno-assembly.html#handbook"><i class="fa fa-check"></i><b>3.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="juno-assembly.html"><a href="juno-assembly.html#requirements-and-preparation-1"><i class="fa fa-check"></i><b>3.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="3.1.2" data-path="juno-assembly.html"><a href="juno-assembly.html#download-the-pipeline"><i class="fa fa-check"></i><b>3.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="3.1.3" data-path="juno-assembly.html"><a href="juno-assembly.html#start-the-analysis.-basics"><i class="fa fa-check"></i><b>3.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="3.1.4" data-path="juno-assembly.html"><a href="juno-assembly.html#output"><i class="fa fa-check"></i><b>3.1.4</b> Output</a></li>
<li class="chapter" data-level="3.1.5" data-path="juno-assembly.html"><a href="juno-assembly.html#troubleshooting-for-this-pipeline"><i class="fa fa-check"></i><b>3.1.5</b> Troubleshooting for this pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="juno-typing.html"><a href="juno-typing.html"><i class="fa fa-check"></i><b>4</b> Juno-typing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="juno-typing.html"><a href="juno-typing.html#handbook-1"><i class="fa fa-check"></i><b>4.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="juno-typing.html"><a href="juno-typing.html#requirements-and-preparation-2"><i class="fa fa-check"></i><b>4.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="4.1.2" data-path="juno-typing.html"><a href="juno-typing.html#download-the-pipeline-1"><i class="fa fa-check"></i><b>4.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="4.1.3" data-path="juno-typing.html"><a href="juno-typing.html#install-conda-environment"><i class="fa fa-check"></i><b>4.1.3</b> Install conda environment</a></li>
<li class="chapter" data-level="4.1.4" data-path="juno-typing.html"><a href="juno-typing.html#start-the-analysis.-basics-1"><i class="fa fa-check"></i><b>4.1.4</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="4.1.5" data-path="juno-typing.html"><a href="juno-typing.html#output-1"><i class="fa fa-check"></i><b>4.1.5</b> Output</a></li>
<li class="chapter" data-level="4.1.6" data-path="juno-typing.html"><a href="juno-typing.html#troubleshooting-for-this-pipeline-1"><i class="fa fa-check"></i><b>4.1.6</b> Troubleshooting for this pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="juno-amr.html"><a href="juno-amr.html"><i class="fa fa-check"></i><b>5</b> Juno-amr</a>
<ul>
<li class="chapter" data-level="5.1" data-path="juno-amr.html"><a href="juno-amr.html#handbook-2"><i class="fa fa-check"></i><b>5.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="juno-amr.html"><a href="juno-amr.html#requirements-and-preparation-3"><i class="fa fa-check"></i><b>5.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="5.1.2" data-path="juno-amr.html"><a href="juno-amr.html#download-the-pipeline-2"><i class="fa fa-check"></i><b>5.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="5.1.3" data-path="juno-amr.html"><a href="juno-amr.html#start-the-analysis.-basics-2"><i class="fa fa-check"></i><b>5.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="5.1.4" data-path="juno-amr.html"><a href="juno-amr.html#output-2"><i class="fa fa-check"></i><b>5.1.4</b> Output</a></li>
<li class="chapter" data-level="5.1.5" data-path="juno-amr.html"><a href="juno-amr.html#troubleshooting-for-this-pipeline-2"><i class="fa fa-check"></i><b>5.1.5</b> Troubleshooting for this pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="juno-annotation.html"><a href="juno-annotation.html"><i class="fa fa-check"></i><b>6</b> Juno-annotation</a>
<ul>
<li class="chapter" data-level="6.1" data-path="juno-annotation.html"><a href="juno-annotation.html#handbook-3"><i class="fa fa-check"></i><b>6.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="juno-annotation.html"><a href="juno-annotation.html#requirements-and-preparation-4"><i class="fa fa-check"></i><b>6.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="6.1.2" data-path="juno-annotation.html"><a href="juno-annotation.html#download-the-pipeline-3"><i class="fa fa-check"></i><b>6.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="6.1.3" data-path="juno-annotation.html"><a href="juno-annotation.html#start-the-analysis.-basics-3"><i class="fa fa-check"></i><b>6.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="6.1.4" data-path="juno-annotation.html"><a href="juno-annotation.html#output-3"><i class="fa fa-check"></i><b>6.1.4</b> Output</a></li>
<li class="chapter" data-level="6.1.5" data-path="juno-annotation.html"><a href="juno-annotation.html#troubleshooting-for-this-pipeline-3"><i class="fa fa-check"></i><b>6.1.5</b> Troubleshooting for this pipeline</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="juno-blast.html"><a href="juno-blast.html"><i class="fa fa-check"></i><b>7</b> Juno_blast</a>
<ul>
<li class="chapter" data-level="7.1" data-path="juno-blast.html"><a href="juno-blast.html#handbook-4"><i class="fa fa-check"></i><b>7.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="juno-blast.html"><a href="juno-blast.html#requirements-and-preparation-5"><i class="fa fa-check"></i><b>7.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="7.1.2" data-path="juno-blast.html"><a href="juno-blast.html#download-the-pipeline-4"><i class="fa fa-check"></i><b>7.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="7.1.3" data-path="juno-blast.html"><a href="juno-blast.html#install-conda-environment-1"><i class="fa fa-check"></i><b>7.1.3</b> Install conda environment</a></li>
<li class="chapter" data-level="7.1.4" data-path="juno-blast.html"><a href="juno-blast.html#start-the-analysis.-basics-4"><i class="fa fa-check"></i><b>7.1.4</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="7.1.5" data-path="juno-blast.html"><a href="juno-blast.html#output-4"><i class="fa fa-check"></i><b>7.1.5</b> Output</a></li>
<li class="chapter" data-level="7.1.6" data-path="juno-blast.html"><a href="juno-blast.html#troubleshooting-for-this-pipeline-4"><i class="fa fa-check"></i><b>7.1.6</b> Troubleshooting for this pipeline</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>III Other scripts/pipelines</b></span></li>
<li class="chapter" data-level="8" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html"><i class="fa fa-check"></i><b>8</b> Low frequency variants in <em>Mycobacterium</em> samples</a>
<ul>
<li class="chapter" data-level="8.1" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#handbook-5"><i class="fa fa-check"></i><b>8.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#requirements-and-preparation-6"><i class="fa fa-check"></i><b>8.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="8.1.2" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#download-the-pipeline-5"><i class="fa fa-check"></i><b>8.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="8.1.3" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#start-the-analysis.-basics-5"><i class="fa fa-check"></i><b>8.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="8.1.4" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#output-5"><i class="fa fa-check"></i><b>8.1.4</b> Output</a></li>
<li class="chapter" data-level="8.1.5" data-path="Myco-lofreq.html"><a href="Myco-lofreq.html#troubleshooting"><i class="fa fa-check"></i><b>8.1.5</b> Troubleshooting</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html"><i class="fa fa-check"></i><b>9</b> Juno-deinterleave</a>
<ul>
<li class="chapter" data-level="9.1" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#handbook-6"><i class="fa fa-check"></i><b>9.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#requirements-and-preparation-7"><i class="fa fa-check"></i><b>9.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="9.1.2" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#download-the-pipeline-6"><i class="fa fa-check"></i><b>9.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="9.1.3" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#start-the-analysis.-basics-6"><i class="fa fa-check"></i><b>9.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="9.1.4" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#output-6"><i class="fa fa-check"></i><b>9.1.4</b> Output</a></li>
<li class="chapter" data-level="9.1.5" data-path="juno-deinterleave.html"><a href="juno-deinterleave.html#extra-options"><i class="fa fa-check"></i><b>9.1.5</b> Extra options</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>IV Microbiome pipelines</b></span></li>
<li class="chapter" data-level="10" data-path="triumph-run.html"><a href="triumph-run.html"><i class="fa fa-check"></i><b>10</b> Triumph: Run pipeline</a>
<ul>
<li class="chapter" data-level="10.1" data-path="triumph-run.html"><a href="triumph-run.html#handbook-7"><i class="fa fa-check"></i><b>10.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="triumph-run.html"><a href="triumph-run.html#requirements-and-preparation-8"><i class="fa fa-check"></i><b>10.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="10.1.2" data-path="triumph-run.html"><a href="triumph-run.html#download-the-pipeline-7"><i class="fa fa-check"></i><b>10.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="10.1.3" data-path="triumph-run.html"><a href="triumph-run.html#start-the-analysis"><i class="fa fa-check"></i><b>10.1.3</b> Start the analysis</a></li>
<li class="chapter" data-level="10.1.4" data-path="triumph-run.html"><a href="triumph-run.html#output-7"><i class="fa fa-check"></i><b>10.1.4</b> Output</a></li>
<li class="chapter" data-level="10.1.5" data-path="triumph-run.html"><a href="triumph-run.html#troubleshooting-1"><i class="fa fa-check"></i><b>10.1.5</b> Troubleshooting</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="11" data-path="triumph-project.html"><a href="triumph-project.html"><i class="fa fa-check"></i><b>11</b> Triumph: Project report</a>
<ul>
<li class="chapter" data-level="11.1" data-path="triumph-project.html"><a href="triumph-project.html#handbook-8"><i class="fa fa-check"></i><b>11.1</b> Handbook</a>
<ul>
<li class="chapter" data-level="11.1.1" data-path="triumph-project.html"><a href="triumph-project.html#requirements-and-preparation-9"><i class="fa fa-check"></i><b>11.1.1</b> Requirements and preparation</a></li>
<li class="chapter" data-level="11.1.2" data-path="triumph-project.html"><a href="triumph-project.html#download-the-pipeline-8"><i class="fa fa-check"></i><b>11.1.2</b> Download the pipeline</a></li>
<li class="chapter" data-level="11.1.3" data-path="triumph-project.html"><a href="triumph-project.html#start-the-analysis.-basics-7"><i class="fa fa-check"></i><b>11.1.3</b> Start the analysis. Basics</a></li>
<li class="chapter" data-level="11.1.4" data-path="triumph-project.html"><a href="triumph-project.html#output-8"><i class="fa fa-check"></i><b>11.1.4</b> Output</a></li>
<li class="chapter" data-level="11.1.5" data-path="triumph-project.html"><a href="triumph-project.html#troubleshooting-2"><i class="fa fa-check"></i><b>11.1.5</b> Troubleshooting</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li>IDS-bioinformatics team - Juno team</li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Bioinformatics Pipelines for Bacteriology (IDS, RIVM)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="juno-assembly" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Juno-assembly pipeline</h1>
<style>
body {
text-align: justify}
</style>
<p><img src="figures/juno_logo/juno_light/juno_assembly_lightbg.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The goal of this pipeline is to generate assemblies from raw fastq files. The input of the pipeline is raw Illumina paired-end data in the form of two fastq files (with extension .fastq, .fastq.gz, .fq or .fq.gz), containing the forward and the reversed reads (’_R1’ and ’_R2’ must be part of the file name, respectively). On the basis of the generated genome assemblies, low quality and contaminated samples can be excluded for downstream analysis.</p>
<p>The pipeline uses the following tools:</p>
<ol style="list-style-type: decimal">
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> (Andrews, 2010) is used to assess the quality of Illumina reads. In this pipeline, it is used before and after trimming/filtering<br />
</li>
<li><a href="https://github.com/OpenGene/fastp">FastP</a> (Chen, Zhou, Chen and Gu, 2018) is used to remove poor quality data and adapter sequences<br />
</li>
<li><a href="https://broadinstitute.github.io/picard/">Picard</a> determines the library fragment lengths<br />
</li>
<li>The reads are assembled into scaffolds by <a href="https://cab.spbu.ru/software/spades/">SPAdes</a> (Bankevich et al., 2012) by means of <em>de novo</em> assembly of the genome. SPAdes uses k-mers for building an initial de Bruijn graph and on following stages it performs graph-theoretical operations to assemble the genome. Kmer sizes of 21, 33, 55, 77 and 99 were used. For <em>de novo</em> assembly, SPAdes ‘isolate’ mode is used<br />
</li>
<li><a href="http://quast.sourceforge.net/">QUAST</a> (Gurevich, Saveliev, Vyahhi, &amp; Tesler, 2013) is used to assess the quality of the filtered scaffolds<br />
</li>
<li>To assess the quality of the microbial genomes, <a href="https://ecogenomics.github.io/CheckM/">CheckM</a> (Parks, Imelfort, Skennerton, Hugenholtz, &amp; Tyson, 2015) is used. CheckM calculates scores for completeness, contamination and strain heterogeneity<br />
</li>
<li><a href="https://jgi.doe.gov/data-and-tools/bbtools/">Bbtools</a> (Bushnell, 2014) is used to generate scaffold alignment metrics<br />
</li>
<li><a href="https://multiqc.info/">MultiQC</a> (Ewels, Magnusson, Lundin, &amp; Käller, 2016) is used to summarize analysis results and quality assessments in a single report for dynamic visualization.<br />
</li>
<li><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> and <a href="http://ccb.jhu.edu/software/bracken/">Bracken</a> for identification of bacterial species.</li>
</ol>
<p><img src="figures/Juno_assembly_DAG.svg" /><!-- --></p>
<div id="handbook" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Handbook</h2>
<div id="requirements-and-preparation-1" class="section level3" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Requirements and preparation</h3>
<p>See the <a href="index.html#general-instructions">General Instructions for all pipelines</a> first.</p>
<ul>
<li>You have different possibilities to run the complete pipeline.
<ul>
<li>Your input directory should contain samples with just ONE genus type. If this is the case, you should tell the pipeline which genus you have (e.g. using the flag <code>--genus Salmonella</code>). Read further for more details.<br />
</li>
<li>You can provide a metadata CSV file that should contain at least two columns: ‘sample’ and ‘genus’ (all in small letters). . If this is not the case DO NOT MODIFY THAT EXCEL FILE! Instead follow the steps mentioned in the Troubleshooting section, particularly in the part “Error saying that my samples cannot be found on the Excel file with the genus list”<br />
</li>
<li>If you do not do any of the above, the pipeline will run just fine. The only difference is that the genome completeness will be calculated based on the genus identified in your sample (the one with most confidence) instead of based on the genus you provided.</li>
</ul></li>
</ul>
</div>
<div id="download-the-pipeline" class="section level3" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> Download the pipeline</h3>
<p><strong>YOU NEED TO DOWNLOAD THE PIPELINE ONCE OR EVERY TIME YOU WANT TO UPDATE IT</strong></p>
<p>Make sure to have followed the <a href="index.html#set-conda">instructions to set up conda</a> before installing any of our pipelines!</p>
<p>Please follow the <a href="index.html#downloading-instructions">instructions to download pipelines</a> from the Juno team of the IDS-bioinformatics group. The Juno-assembly pipeline can be found in this link: <a href="https://github.com/AleSR13/Juno_pipeline">Juno-assembly pipeline</a>.</p>
<p>After downloading the pipeline, it needs to be installed. To install the pipeline, first enter the folder where you downloaded the pipeline and run the installation script:</p>
<pre><code>cd /mnt/scratch_dir/&lt;my_folder&gt;/Juno_pipeline
bash install_juno_assembly.sh</code></pre>
</div>
<div id="start-the-analysis.-basics" class="section level3" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Start the analysis. Basics</h3>
<ol style="list-style-type: decimal">
<li>Open a terminal. (Applications &gt; terminal).<br />
</li>
<li>Enter the folder of the pipeline using:</li>
</ol>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="juno-assembly.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/Juno_pipeline</span></code></pre></div>
<p>MAKE SURE YOU HAVE RUN THE INSTALLATION INSTRUCTIONS BEFORE USING THE PIPELINE FOR THE FIRST TIME OR AFTER AN UPDATE!</p>
<ol start="3" style="list-style-type: decimal">
<li>Activate the ‘environment’ that has the necessary software for the pipeline to work:</li>
</ol>
<pre><code>conda activate juno_assembly</code></pre>
<p>If you run in trouble please see the <a href="general-troubleshooting.html#problems-conda-activate">troubleshooting section for conda activate</a>.</p>
<ol start="4" style="list-style-type: decimal">
<li>Run the pipeline</li>
</ol>
<p>If all your samples have the same genus, for instance, <em>Salmonella</em>, you run it like this:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="juno-assembly.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> juno_assembly.py <span class="at">-i</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_data<span class="op">&gt;</span>/ <span class="at">--genus</span> Salmonella</span></code></pre></div>
<ul>
<li>Note that the genus should be ONE word. Do not put any species names!</li>
</ul>
<p>If you want to provide a metadata file do this:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="juno-assembly.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> juno_assembly.py <span class="at">-i</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_data<span class="op">&gt;</span>/ <span class="at">--metadata</span> <span class="op">&lt;</span>path_to_file<span class="op">&gt;</span>.csv</span></code></pre></div>
<p>If you do not want to provide the genus yourself, you can simply do :</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="juno-assembly.html#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> juno_assembly.py <span class="at">-i</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_data<span class="op">&gt;</span>/</span></code></pre></div>
<p>Please read the section <a href="index.html#what-to-expect">What to expect while running a Juno pipeline</a></p>
<p>See the section <a href="general-troubleshooting.html#general-troubleshooting">General Troubleshooting</a> for any problems you may encounter.</p>
<p><strong>Note:</strong> Do not keep all your data (including results) on the scratch_dir partition. You are allowed to keep 400GB max and with sequencing data, this can get full quite fast.</p>
<ol start="5" style="list-style-type: decimal">
<li>Optionally, deactivate the juno_assembly environment.</li>
</ol>
<pre><code>conda deactivate</code></pre>
</div>
<div id="output" class="section level3" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> Output</h3>
<p>A folder called <code>output/</code>, inside the folder of the pipeline, will be created. This folder will contain all the results and logging files of your analysis. There will be one folder per result (<code>qc_raw_fastq</code>, <code>clean_fastq</code>, <code>qc_clean_fastq</code>, <code>de_novo_assembly</code>, <code>de_novo_assembly_filtered</code>, <code>qc_de_novo_assembly</code>, <code>identify_species</code>). Please refer to the manuals of every tool to interpret the results. .</p>
<p><em>Note:</em> If you want your output to be stored in a folder with a different name or location, you can use the option <code>-o</code> (‘o’ from output)</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="juno-assembly.html#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> juno_assembly.py <span class="at">-i</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_data<span class="op">&gt;</span>/ <span class="at">-o</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_results<span class="op">&gt;</span>/</span></code></pre></div>
<p>Another very important output from the pipeline are the logging files and audit trail that contain information of the software versions used, the parameters used, the error messages, etc. They could be important for you if you want to publish or reproduce the analysis at a later time point and also to get help from the bioinformatics team if you were to run into trouble with the pipeline. Please read about these files <a href="index.html#general-output">here</a>.</p>
</div>
<div id="troubleshooting-for-this-pipeline" class="section level3" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> Troubleshooting for this pipeline</h3>
<p><strong>Please read first the <a href="general-troubleshooting.html#general-troubleshooting">General Troubleshooting</a> section!</strong></p>
<div id="error-saying-that-the-genus-supplied-was-not-recognized-by-checkm" class="section level4" number="3.1.5.1">
<h4><span class="header-section-number">3.1.5.1</span> Error saying that the genus supplied was not recognized by CheckM</h4>
<p>If you get this message:</p>
<pre><code>The genus &lt;your_provided_genus&gt; is not supported. </code></pre>
<p>it means that although you provided the genus information, it is probably not supported by the tool <code>CheckM</code>. First check that you have no spelling mistakes in the genus name of your sample(s) or that you used the correct case (first letter capital and the rest on small letters). If you are sure that your spelling is correct, it may be that CheckM does not have the genus you supplied on its database. As the message says, if you are not sure on how to spell the genus name or which genus is accepted, please check the list of accepted genera by typing.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb26-1"><a href="juno-assembly.html#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> juno_assembly.py <span class="at">-i</span> /mnt/scratch_dir/<span class="op">&lt;</span>my_folder<span class="op">&gt;</span>/<span class="op">&lt;</span>my_data<span class="op">&gt;</span>/ <span class="at">--help-genera</span></span></code></pre></div>
</div>
<div id="other-problems-or-failing-rules-1" class="section level4" number="3.1.5.2">
<h4><span class="header-section-number">3.1.5.2</span> Other problems or failing rules</h4>
<p>The Juno-assembly pipeline is still in development which means that sometimes the process can fail.</p>
<p>Before contacting for help, try these two steps:</p>
<ol style="list-style-type: decimal">
<li><p>Re-run the pipeline again and see if the process continues. If it does, please keep re-running the pipeline until your analysis is finished or there is no longer progress. In this case, send an email after the pipeline is finished so I can troubleshoot the problem.</p></li>
<li><p><a href="https://github.com/AleSR13/Juno_pipeline">Download</a> the pipeline again and start from the beginning of this <a href="juno-assembly.html#juno-assembly">handbook</a>. Sometimes there is an issue that has been resolved in newer versions of the pipeline.</p></li>
</ol>
<p>If the pipeline still fails after these two steps, please inform me about the problem. Send an <a href="mailto:alejandra.hernandez.segura@rivm.nl">e-mail</a> with the following content:</p>
<ul>
<li>The log and error files that can be found in the output folder</li>
<li>The path to your input directory</li>
<li>The path to where the pipeline is installed</li>
</ul>
<p><strong>Note: </strong> I cannot help you without this information, if information is missing there will be a delay in troubleshooting the problem.</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="general-troubleshooting.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="juno-typing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": null,
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection",
"scroll_highlight": true
},
"toolbar": {
"position": "fixed"
},
"search": true,
"info": true
});
});
</script>

</body>

</html>
